src/H5Fint.c

```diff
@@ -1850,56 +1860,54 @@ H5F_open(const char *name, unsigned flags, hid_t fcpl_id, hid_t fapl_id)
-     * tentative file opens, surrounding the file open call with the
-     * H5E_BEGIN/END_TRY macros is an explicit instruction to the file driver
-     * not to display errors. If the tentative file open call fails, another
-     * attempt at opening the file will be made without error output being
-     * suppressed.
-     *
-     * However, if stripping away the H5F_ACC_CREAT flag and others left us
-     * with the same file access flags as before, then we will skip this
-     * tentative file open and only make a single attempt at opening the file.
-     * In this case, we don't want to suppress error output since the underlying
-     * file driver might provide more details on why the file open failed.
-     */
-    if (tent_flags != flags) {
-        /* Make tentative attempt to open file */
-        H5E_BEGIN_TRY
-        {
-            lf = H5FD_open(name, tent_flags, fapl_id, HADDR_UNDEF);
+        /*
+         * When performing a tentative open of a file where we have stripped
+         * away flags such as H5F_ACC_CREAT from the specified file access
+         * flags, use 'try open' operation to avoid pushing error messages
+         * on the error stack since there is an expectation that the tentative
+         * open might fail.
+         *
+         * If the tentative file open call fails, another attempt at opening
+         * the file will be made without error output being suppressed.
+         *
+         * However, if stripping away the H5F_ACC_CREAT flag and others left
+         * us with the same file access flags as before, then we will skip
+         * this tentative file open and only make a single attempt at opening
+         * the file.  In this case, we don't want to suppress errors since
+         * the underlying file driver might provide more details on why the
+         * file open failed.
+         */
+        if (tent_flags != flags) {
+            /* Make tentative attempt to open file */
+            if (H5FD_try_open(&lf, name, tent_flags, fapl_id, HADDR_UNDEF) < 0)
+                HGOTO_ERROR(H5E_FILE, H5E_CANTOPENFILE, FAIL, "can't try opening file");
+
+            /* If the tentative open failed, reset the file access flags,
+             * then make another attempt at opening the file.
+             */
+            if (NULL == lf)
+                tent_flags = flags;
         }
-        H5E_END_TRY
     }
+    else
+        tent_flags = flags;
 
     /*
      * If a tentative attempt to open the file wasn't necessary, attempt
      * to open the file now. Otherwise, if the tentative open failed, clear
      * the error stack and reset the file access flags, then make another
      * attempt at opening the file.
      */
-    if ((tent_flags == flags) || (lf == NULL)) {
-        if (tent_flags != flags) {
-            H5E_clear_stack();
-            tent_flags = flags;
-        }
+    if (NULL == lf) {
+        if (H5FD_try_open(&lf, name, tent_flags, fapl_id, HADDR_UNDEF) < 0)
```

Comment:
This is a case where I'd think rather than switching to a try_open we'd want to keep the same behavior of just a simple open since this is the final open attempt and we're interested in knowing if something fails.

Reply:
No, I believe that this is correct now.  This is in H5F_try_open and the code that calls this routine is built around trying the open, so that errors are only generated when there's no more opportunities for trying to open the file.   It's not an error to try to open the file and not succeed, the file pointer is just set to NULL, and the calling routine can determine what action it wants to take, which could involve trying again.

For example,  H5F_try_open is called from H5F__efc_try_open_file, which is called by H5F__efc_try_open, which is called in H5F_prefix_try_open() several times, to attempt opening the file with various paths, etc.

The logic for these new "try open" and "open" calls all follow the same pattern, to avoid clearing the error stack unnecessarily.

Reply:
The previous code for `H5F_open`, for example, was clearing the error stack after the speculative file open fails, which is perfectly reasonable, but it wasn't clearing the error stack after the final file open attempt fails so that any errors from the VFD level can be captured. Since `H5F_open` is changed here to just call `H5F_try_open` once, which does a `H5FD_try_open` for both file open attempts, it seems like the error stack will be paused for both file open attempts and code that calls `H5F_open` won't be able to capture errors from the VFD used.

Reply:
@jhendersonHDF - I updated to make the changes we talked about yesterday.  How's this look now?